File: 11127/main.c
#include <stdio.h>

int is_triple_free(int binary, int bit)
{
   int t[2];
   int mask = (1 << bit) - 1;

   t[0] = binary & mask;
   binary >>= bit;
   t[1] = binary & mask;
   binary >>= bit;
   binary &= mask;

   return (t[0] == t[1] && t[1] == binary);
}

int triple_count(char* input, int binary, int n)
{
   int bit;

   for (bit = n / 3; bit > 0; bit--) {
      if (is_triple_free(binary, bit))
         return 0;
   }

   if (*input == '\0')
      return 1;

   if (*input == '*')
      return triple_count((input + 1), (binary << 1), n + 1) + triple_count((input + 1), (binary << 1) + 1, n + 1);

   if (*input == '0')
      return triple_count((input + 1), (binary << 1), n + 1);
   else
      return triple_count((input + 1), (binary << 1) + 1, n + 1);
}

int main(int argc, char* argv[])
{
   int n, test_case = 1;
   char input[32];

   while (scanf("%d", &n) == 1 && n != 0) {
      scanf("%s", input);
      printf("Case %d: %d\n", test_case, triple_count(input, 0, 0));
      ++test_case;
   }

   return 0;
}

File: 11127/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String args[]) throws IOException {
      int testCase = 1;
      String line;
      StringBuilder answer = new StringBuilder();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         if (line.charAt(0) == '0') break;

         StringTokenizer tokenizer = new StringTokenizer(line);
         String n = tokenizer.nextToken();
         String binaryString = tokenizer.nextToken();

         answer.append("Case ");
         answer.append(Integer.toString(testCase));
         answer.append(": ");
         answer.append(Integer.toString(countTriple(binaryString, 0, 0)));
         answer.append('\n');
         ++testCase;
      }
      System.out.print(answer);
   }

   public static int countTriple(String binaryString, int binary, int n) {
      for (int bit = n / 3; bit > 0; bit--) {
         if (isTripleFree(binary, bit))
         return 0;
      }

      if (n == binaryString.length())
         return 1;

      if (binaryString.charAt(n) == '*')
         return countTriple(binaryString, binary << 1, n + 1) + countTriple(binaryString, (binary << 1) + 1, n + 1);

      if (binaryString.charAt(n) == '0')
         return countTriple(binaryString, binary << 1, n + 1);
      else
         return countTriple(binaryString, (binary << 1) + 1, n + 1);
   }

   public static boolean isTripleFree(int binary, int bit) {
      int[] t = new int[2];
      int mask = (1 << bit) - 1;

      t[0] = binary & mask;
      binary >>= bit;
      t[1] = binary & mask;
      binary >>= bit;
      binary &= mask;

      return (t[0] == t[1] && t[1] == binary);
   }
}
